whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -62,54 +62,77 @@ public RdapRelationService(final Ipv4Tree ip4Tree,
                                final Ipv6DomainTree ipv6DomainTree,
                                final RpslObjectDao rpslObjectDao,
                                final RdapRequestValidator rdapRequestValidator,
-                               final RdapQueryHandler rdapQueryHandler) {
+                               final RdapQueryHandler rdapQueryHandler,
+                               final RdapObjectMapper rdapObjectMapper,
+                               final RdapLookupService rdapLookupService) {
         this.ip4Tree = ip4Tree;
         this.ip6Tree = ip6Tree;
         this.ipv4DomainTree = ipv4DomainTree;
         this.ipv6DomainTree = ipv6DomainTree;
         this.rpslObjectDao = rpslObjectDao;
         this.rdapRequestValidator = rdapRequestValidator;
         this.rdapQueryHandler = rdapQueryHandler;
+        this.rdapObjectMapper = rdapObjectMapper;
+        this.rdapLookupService = rdapLookupService;
     }
 
-    protected List<RpslObject> handleRelationQuery(final HttpServletRequest request, final RdapRequestType requestType,
-                                                 final RelationType relationType, final String key) {
+    protected Object handleRelationQuery(final HttpServletRequest request,
+                                         final Set<ObjectType> objectTypes, final RdapRequestType requestType,
+                                         final RelationType relationType, final String key,
+                                         final String requestUrl,
+                                         final int maxResultSize) {
         final List<RpslObject> rpslObjects;
+        final boolean shouldReturnLookup = relationType.equals(RelationType.UP) || relationType.equals(RelationType.TOP);
         switch (requestType) {
             case AUTNUMS -> throw new RdapException("501 Not Implemented", "Relation queries not allowed for autnum", HttpStatus.NOT_IMPLEMENTED_501);
             case DOMAINS -> {
                 rdapRequestValidator.validateDomain(key);
-                final List<String> relatedPkeys = getDomainsByRelationType(key, relationType);
+                final List<IpEntry> domainEntries = getDomainsEntriesByRelationType(key, relationType);
+
+                if (shouldReturnLookup){
+                    final IpEntry ipEntry = domainEntries.getFirst();
+                    final RpslObject domainObject = rpslObjectDao.getById(ipEntry.getObjectId());
+                    return rdapLookupService.lookupForDomain(request, domainObject.getKey().toString(), ipEntry.getKey().toString());
```

Comment:
LookupForDomain is again hitting  database to fetch domain Object ? Can you avoid multiple calls ? May eb overload the function?